tests/TagTest.php

```diff
@@ -44,4 +44,152 @@ public function testVoidTag()
+    {
+        $tag = new Tag('img');
+        $tag->setTextContent('text_content');
+    }
+
+    /**
+     * @expectedException        \LogicException
+     * @expectedExceptionMessage Void elements cannot have any contents.
+     */
+    public function testSetHtmlContentWithVoidElementShouldReturnLogicException()
+    {
+        $tag = new Tag('img');
+        $tag->setHtmlContent('<p>html_content</p>');
+    }
+
+    /**
+     * @expectedException        \LogicException
+     * @expectedExceptionMessage Void elements cannot have any contents.
+     */
+    public function testAppendTextContentWithVoidElementShouldReturnLogicException()
+    {
+        $tag = new Tag('img');
+        $tag->appendTextContent('this_is_appended_text_content');
+    }
+
+    public function testAppendTextContent()
+    {
+        $tag = new Tag('p');
+        $appendTextContent = 'this_is_appended_text_content';
+        $tag->appendTextContent($appendTextContent);
+
+        $this->assertSame('<p>' . $appendTextContent . '</p>', $tag->render());
+    }
+
+    /**
+     * @expectedException        \LogicException
+     * @expectedExceptionMessage Void elements cannot have any contents.
+     */
+    public function testAppendHtmlContentWithVoidElementShouldReturnLogicException()
+    {
+        $tag = new Tag('img');
+        $tag->appendHtmlContent('<span>this_is_appended_html_content</span>');
+    }
+
+    public function testAppendHtmlContent()
+    {
+        $tag = new Tag('p');
+        $appendHtmlContent = '<span>this_is_appended_text_content</span>';
+        $tag->appendHtmlContent($appendHtmlContent);
+
+        $this->assertSame('<p>' . $appendHtmlContent . '</p>', $tag->render());
+    }
+
+    /**
+     * @expectedException        \LogicException
+     * @expectedExceptionMessage Void elements cannot have any contents.
+     */
+    public function testAppendWithVoidElementShouldReturnLogicException()
+    {
+        $tag = new Tag('img');
+        $tag->append(new Tag('img'));
+    }
+
+    public function testIsEmptyShouldReturnTrue()
+    {
+        $tag = new Tag('p');
+        $tag->appendTextContent('append_text_content');
+        $tag->empty();
+
+        $this->assertTrue($tag->isEmpty());
+    }
+
+    public function testIsEmptyShouldReturnFale()
+    {
+        $tag = new Tag('p');
+        $tag->appendTextContent('append_text_content');
+
+        $this->assertFalse($tag->isEmpty());
+    }
+
+    /**
+     * @expectedException        \LogicException
+     * @expectedExceptionMessage Void elements do not have a closing tag.
+     */
+    public function testRenderClosingTag()
+    {
+        $tag = new Tag('img');
+        $tag->renderClosingTag();
+    }
+
+    public function testTagInstanceShouldReturnRenderedString()
+    {
+        $tag = new Tag('p');
+        $appendTextContent = 'append_text_content';
+        $tag->appendTextContent($appendTextContent);
+        $tag->render();
```

Comment:
You don't need to call `render()` before casting to string! `__toString()` returns `render()`.
You should actually return the return value of both `render()` and `(string) $tag` in this method.